### PR TITLE
Add optional weighted image sampler

### PR DESCRIPTION
**Pull request by eharjes**

### CODE REVIEW

File: utils/dataloaders.py [hunk 1]
```diff
@@ -126,20 +128,33 @@ def create_dataloader(path,
...
-    return loader(dataset,
-                  batch_size=batch_size,
-                  shuffle=shuffle and sampler is None,
-                  num_workers=nw,
-                  sampler=sampler,
-                  pin_memory=True,
-                  collate_fn=LoadImagesAndLabels.collate_fn4 if quad else LoadImagesAndLabels.collate_fn), dataset
+
+    if balanced:
+        sampler = WeightedRandomSampler(dataset._weights, len(dataset))
```
Comment (jootten) [inline utils/dataloaders.py:139]: Could you simplify this? For example, by using the `balanced` kwarg only to change the sampler. Also, I think we don't need the `BatchSampler`, instead we can directly use the sampler.

File: utils/dataloaders.py [hunk 2]
```diff
@@ -876,6 +895,18 @@ def collate_fn4(batch):
 
         return torch.stack(im4, 0), torch.cat(label4, 0), path4, shapes4
 
+    def _compute_occurence_weights(self):
+        all_class_labels = np.concatenate([lbl[:, 0] for lbl in self.labels])
+        class_indices = np.unique(all_class_labels)
+        all_class_labels = all_class_labels.tolist()
+        inverse_class_frequencies = {idx: 1 / all_class_labels.count(idx) for idx in class_indices}
```
Comment (jootten) [inline utils/dataloaders.py:902]: Let's use `np.bincount` to count the occurrences, this should be more performant, and we don't need to convert `all_classes_labels` to a list anymore.

https://numpy.org/doc/stable/reference/generated/numpy.bincount.html

File: utils/dataloaders.py [hunk 3]
```diff
@@ -888,12 +888,11 @@ def collate_fn4(batch):
 
     def _compute_occurence_weights(self):
         all_class_labels = np.concatenate([lbl[:, 0] for lbl in self.labels])
```
Comment (jootten) [inline utils/dataloaders.py:890]: You convert this to `np.int64` below twice. It should suffice to do it once here. Also, please use `np.uint16` because the class indices are all positive and won't be higher than 32768. This will save some space.

File: utils/dataloaders.py [hunk 4]
```diff
@@ -888,12 +888,11 @@ def collate_fn4(batch):
...
-        class_indices = np.unique(all_class_labels)
-        all_class_labels = all_class_labels.tolist()
-        inverse_class_frequencies = {idx: 1 / all_class_labels.count(idx) for idx in class_indices}
+        class_frequencies = np.bincount(all_class_labels.astype(np.int64))
         weights = []
         for lbl in self.labels:
-            weights.append(np.prod([inverse_class_frequencies[idx] for idx in lbl[:,0]]))
+            weights.append(np.prod([1/class_frequencies[idx.astype(np.int64)] for idx in lbl[:,0]]))
```
Comment (jootten) [inline utils/dataloaders.py:894]: Python convention: Usually, you put spaces between binary operators and after commas.
```python
            weights.append(np.prod([1 / class_frequencies[idx] for idx in lbl[:, 0]]))
```

File: utils/dataloaders.py [hunk 5]
```diff
@@ -888,12 +888,11 @@ def collate_fn4(batch):
...
+        print(weights)
```
Comment (jootten) [inline utils/dataloaders.py:895]: Remove print statements :D
```suggestion
```

File: utils/dataloaders.py (hunk 3 above)
Comment (eharjes) [reply utils/dataloaders.py:890]: I think you still need it twice. One time for all_class_labels to calculate the class frequencies and one time for the class indices when iterating over self.labels.
